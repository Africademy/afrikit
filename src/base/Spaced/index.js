// @flow
import React from 'react';
import styled from 'styled-components';
import { space } from '../../theme/space';

const classnames = (...args) => args.join(' ');
const getClassName = el => (el.props && el.props.className) || '';

export class StyledChildren extends React.Component {
  render() {
    // eslint-disable-next-line
    const { className, children } = this.props;
    const styledChildren = React.Children.toArray(children).map(child =>
      React.cloneElement(child, {
        className: classnames(getClassName(child), className),
      }),
    );

    return <React.Fragment>{styledChildren}</React.Fragment>;
  }
}

const Spaced = styled(StyledChildren)([], space);

Spaced.propTypes = space.propTypes;

export default Spaced;
