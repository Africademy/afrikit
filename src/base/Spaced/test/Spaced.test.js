// @flow
import React from 'react';
import expect from 'expect';
import { create as render } from 'react-test-renderer';
import Spaced from '../index';

const renderJSON = el => render(el).toJSON();

describe('Spaced Base Component', () => {
  test('renders', () => {
    const json = renderJSON(<Spaced />);
    expect(json).toMatchSnapshot();
  });

  test('renders children', () => {
    const json = renderJSON(
      <Spaced>
        <div>Hello</div>
        <h2>hi</h2>
      </Spaced>,
    );
    expect(json).toMatchSnapshot();
  });

  test('adds classNames to children', () => {
    const json = renderJSON(
      <Spaced mx={2}>
        <div>Hello</div>
        <h2>hi</h2>
      </Spaced>,
    );
    const { className } = json[0].props;

    expect(json[0].props.className.length > 0).toBeTruthy();
    expect(json[1].props.className).toEqual(className);
  });

  test('merges with existing child classNames', () => {
    const json = renderJSON(
      <Spaced mx={2}>
        <div className="beep">Hello</div>
        <h2>hi</h2>
      </Spaced>,
    );
    const { className } = json[0].props;
    expect(className).toMatch(/^beep\s/);
  });
});
