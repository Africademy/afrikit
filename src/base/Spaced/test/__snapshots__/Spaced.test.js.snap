// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spaced Base Component renders 1`] = `null`;

exports[`Spaced Base Component renders children 1`] = `
Array [
  <div
    className=" sc-bdVaJa bDWFJH"
  >
    Hello
  </div>,
  <h2
    className=" sc-bdVaJa bDWFJH"
  >
    hi
  </h2>,
]
`;
